### PR TITLE
Added private link option to the exfiltration protection example

### PR DESCRIPTION
**Pull request by mwojtyczka**
The example currently doesn't have a private link setup included as indicated in the architecture diagram. The PR adds an optional private link along with other changes based on field experience.

Changes:
* Added missing private link setup. Added variable to enabled/disabled the private link. By default disabled since it requires the Databricks Enterprise tier so won't work for all customers.
* Added ready configuration for each region (webapp, SCC relay, hive metastore, control plan). The configuration can still be provided explicitly as before. This greatly simplifies the setup for the customers.
* Updated Readme
* Added output.tf so that users don't have to do it
* Upgraded provider (databricks, aws) versions
* Whitelisted repositories for maven and spark packages
* Updated aws services setup as per the latest aws provider version
* Refactored in places and provided better comments
* Deployment tested with and without private link enabled

### CODE REVIEW

File: modules/aws-exfiltration-protection/versions.tf [hunk 1]
```diff
@@ -3,12 +3,12 @@ terraform {
...
     databricks = {
       source  = "databricks/databricks"
-      version = ">=0.5.1"
+      version = ">=1.27.0"
     }
 
     aws = {
       source  = "hashicorp/aws"
-      version = "~> 4.58.0"
+      version = "~> 5.20.1"
```
Comment (alexott) [inline modules/aws-exfiltration-protection/versions.tf:11]: We need to remove versions from the modules

**Review by alexott — COMMENTED** [review]
@hwang-db @yessawab can you look as well?

File: examples/aws-exfiltration-protection/README.md [hunk 2]
```diff
@@ -20,7 +23,8 @@ This template provides an example deployment of AWS Databricks E2 workspace with
...
     * TF_VAR_databricks_account_id, set to the value of the ID of your Databricks account. You can find this value in the corner of your Databricks account console.
-5. Add a `output.tf` file.
-6. (Optional) Configure your [remote backend](https://developer.hashicorp.com/terraform/language/settings/backends/s3)
-7. Run `terraform init` to initialize terraform and get provider ready.
-8. Run `terraform apply` to create the resources.
\ No newline at end of file
+5. (Optional) Configure your [remote backend](https://developer.hashicorp.com/terraform/language/settings/backends/s3)
+6. Run `terraform init` to initialize terraform and get provider ready.
+7. Run `terraform plan -var-file=terraform.tfvars` to validate and preview the deployment.
+8. Run `terraform apply -var-file=terraform.tfvars` to create the resources.
```
[the remark this answers is not in the document]
Comment (mwojtyczka) [reply examples/aws-exfiltration-protection/README.md]: removed

File: examples/aws-exfiltration-protection/main.tf [hunk 3]
```diff
@@ -1,6 +1,13 @@
 module "aws-exfiltration-protection" {
-  source                      = "github.com/databricks/terraform-databricks-examples/modules/aws-exfiltration-protection"
+  source                      = "../../modules/aws-exfiltration-protection"
```
[the remark this answers is not in the document]
Comment (mwojtyczka) [reply examples/aws-exfiltration-protection/main.tf]: I reverted this change to be compatible with the rest of the examples. 

It depends on how this is meant to be used. If the modules should not be touched after cloning then it's best.
But if you want to clone and make some modifications you would need to first update the source. Most people will likely want to adjust the module so in that case the local source is better.

File: examples/aws-exfiltration-protection/variables.tf [hunk 4]
```diff
@@ -16,31 +16,20 @@ variable "region" {
...
-variable "db_rds" {
-  default = "mdv2llxgl8lou0.ceptxxgorjrc.eu-central-1.rds.amazonaws.com"
-}
-
-variable "db_control_plane" {
-  default = "18.159.44.32/28"
+  default = [
+    ".pypi.org", ".pythonhosted.org",   # python packages
+    ".cran.r-project.org",              # R packages
+    ".maven.apache.org", ".maven.org",  # maven artifacts
```
[the remark this answers is not in the document]
Comment (mwojtyczka) [reply examples/aws-exfiltration-protection/variables.tf]: I've tested again. It's not needed. I removed it.

File: examples/aws-exfiltration-protection/variables.tf [hunk 5]
```diff
@@ -16,31 +16,20 @@ variable "region" {
...
+    ".spark-packages.org",              # spark packages
+  ]
 }
 
 variable "prefix" {
   default = "demo"
+}
+
+variable "enable_private_link" {
+  default = false
```
[the remark this answers is not in the document]
Comment (mwojtyczka) [reply examples/aws-exfiltration-protection/variables.tf:49]: I added types to all vars. Btw. it's missing in most of the examples we have.

File: modules/aws-exfiltration-protection/README.md [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# Provisioning AWS Databricks E2 with a Hub & Spoke firewall for data exfiltration protection
+# Provisioning AWS Databricks workspace with a Hub & Spoke firewall for data exfiltration protection
 
 This template provides an example deployment of AWS Databricks E2 workspace with a Hub & Spoke firewall for data exfiltration protection. Details are described in [Data Exfiltration Protection With Databricks on AWS](https://www.databricks.com/blog/2021/02/02/data-exfiltration-protection-with-databricks-on-aws.html).
```
[the remark this answers is not in the document]
Comment (mwojtyczka) [reply modules/aws-exfiltration-protection/README.md:3]: I added E2 and made it consistent.

File: modules/aws-exfiltration-protection/README.md [hunk 7]
```diff
@@ -16,12 +16,15 @@ Resources to be created:
 * S3 Root bucket
 * Cross-account IAM role
 * Databricks E2 workspace
+* Private link between clusters on the data plane and core services on the control plane
```
[the remark this answers is not in the document]
Comment (mwojtyczka) [reply modules/aws-exfiltration-protection/README.md]: Added

File: modules/aws-exfiltration-protection/versions.tf (hunk 1 above)
Comment (mwojtyczka) [reply modules/aws-exfiltration-protection/versions.tf:11]: I followed all other examples. Versions are included in all of them.

File: examples/aws-exfiltration-protection/main.tf [hunk 8]
```diff
@@ -3,10 +3,16 @@ module "aws-exfiltration-protection" {
...
 
-resource "random_string" "naming" {
-  special = false
-  upper   = false
-  length  = 6
+  resource "random_string" "naming" {
+    special = false
+    upper   = false
+    length  = 6
 }
```
Comment (alexott) [inline examples/aws-exfiltration-protection/main.tf:18]: I think that we have a missing `}` for the module...

File: examples/aws-exfiltration-protection/variables.tf [hunk 9]
```diff
@@ -25,39 +30,25 @@ variable "hub_cidr_block" {
 }
 
 variable "region" {
-  default     = "eu-central-1"
-  type        = string
+  type    = string
+  default = "eu-central-1"
   description = "AWS region to deploy to"
```
Comment (alexott) [inline examples/aws-exfiltration-protection/variables.tf:35]: Let's format the code...

File: modules/aws-exfiltration-protection/providers.tf [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+}
+
+// initialize provider in "MWS" mode to provision new workspace
+provider "databricks" {
+  alias      = "mws"
+  host       = "https://accounts.cloud.databricks.com"
+  account_id = var.databricks_account_id
+  username   = var.databricks_account_username
+  password   = var.databricks_account_password
+}
```
Comment (alexott) [inline modules/aws-exfiltration-protection/providers.tf:12]: It's better not to do this in the modules. Providers is recommended to pass explicitly configured.

https://developer.hashicorp.com/terraform/language/meta-arguments/module-providers

**Review by alexott — CHANGES_REQUESTED** [review]

File: examples/aws-exfiltration-protection/main.tf (hunk 8 above)
Comment (mwojtyczka) [reply examples/aws-exfiltration-protection/main.tf:18]: corrected

File: examples/aws-exfiltration-protection/variables.tf (hunk 9 above)
Comment (mwojtyczka) [reply examples/aws-exfiltration-protection/variables.tf:35]: done

File: modules/aws-exfiltration-protection/providers.tf (hunk 10 above)
Comment (mwojtyczka) [reply modules/aws-exfiltration-protection/providers.tf]: Moved providers definition from the module to the examples

File: modules/aws-exfiltration-protection/workspace.tf [hunk 11]
```diff
@@ -1,14 +1,23 @@
 resource "databricks_mws_networks" "this" {
-  provider           = databricks.mws
```
Comment (alexott) [inline modules/aws-exfiltration-protection/workspace.tf:2]: why do we need this? Default provider should work

File: modules/aws-exfiltration-protection/variables.tf [hunk 12]
```diff
@@ -3,60 +3,90 @@ variable "databricks_account_id" {
...
 }
 
-variable "tags" {
-  default     = {}
-  type        = map(string)
-  description = "Optional tags to add to created resources"
+resource "random_string" "naming" {
+  special = false
+  upper   = false
+  length  = 6
```
Comment (alexott) [inline modules/aws-exfiltration-protection/variables.tf:9]: Usually `variables.tf` is used only for variables declaration...

**Review by alexott — APPROVED** [review]
looks good, small comment

File: modules/aws-exfiltration-protection/variables.tf (hunk 12 above)
Comment (mwojtyczka) [reply modules/aws-exfiltration-protection/variables.tf:9]: corrected

File: modules/aws-exfiltration-protection/workspace.tf (hunk 11 above)
Comment (mwojtyczka) [reply modules/aws-exfiltration-protection/workspace.tf:2]: removed provider to use the default

**Review by alexott — APPROVED** [review]

**Pull request merged**